src/prommis/nanofiltration/costing/diafiltration.py

```diff
@@ -0,0 +1,713 @@
+C_Li_diaf = 0.1 * pyo.units.kg / pyo.units.m**3
+C_Co_diaf = 0.2 * pyo.units.kg / pyo.units.m**3
+
+
+def main():
+    m = build_and_init_model()
+
+    (
+        R_Li,
+        R_Co,
+        P_Li,
+        P_Co,
+        membrane_length,
+        feed_flow_rate,
+        diafiltrate_flow_rate,
+        permeate_flow_rate,
+    ) = get_values(m)
+
+    add_costing(
+        m,
+        membrane_length,
+        membrane_width=w,
+        water_flux=Jw,
+        feed_flow=feed_flow_rate,
+        perm_flow=permeate_flow_rate,
+        inlet_pressure=diafiltrate_inlet_pressure,
+        outlet_pressure=diafiltrate_outlet_pressure,
+        diaf_flow=diafiltrate_flow_rate,
+    )
+
+    # print(f"DOF = {degrees_of_freedom(m)}")
+    # dt = DiagnosticsToolbox(m)
+    # dt.report_structural_issues()
+    # dt.display_components_with_inconsistent_units()
+    # dt.display_potential_evaluation_errors()
+    # dt.report_numerical_issues()
+
+    unfix_variables(m)
+    add_constraints(m)
+    add_objective(m)
+    print(f"DOF = {degrees_of_freedom(m)}")
+    solve_model(m)
+    print_information(m)
+    # dt.report_structural_issues()
+    # dt.report_numerical_issues()
+
+
+def build_and_init_model():
+    m = pyo.ConcreteModel()
+    m.fs = FlowsheetBlock(dynamic=False)
+    m.fs.properties = LiCoParameters()
+
+    m.fs.stage1 = MSContactor(
+        number_of_finite_elements=10,
+        streams={
+            "retentate": {
+                "property_package": m.fs.properties,
+                "has_energy_balance": False,
+                "has_pressure_balance": False,
+            },
+            "permeate": {
+                "property_package": m.fs.properties,
+                "has_feed": False,
+                "has_energy_balance": False,
+                "has_pressure_balance": False,
+            },
+        },
+    )
+
+    # Add mass transfer variables and constraints
+    m.fs.solutes = pyo.Set(initialize=["Li", "Co"])
+
+    m.fs.sieving_coefficient = pyo.Var(
+        m.fs.solutes,
+        units=pyo.units.dimensionless,
+    )
+    m.fs.sieving_coefficient["Li"].fix(1.3)
+    m.fs.sieving_coefficient["Co"].fix(0.5)
+
+    m.fs.stage1.length = pyo.Var(units=pyo.units.m)
+
+    # Start by initializing with a short length
+    # Too long and we deplete solvent due to lack of recycles
+    m.fs.stage1.length.fix(10)
+
+    solvent_rule, solute_rule, stage3_solute_rule = create_rules(m)
+
+    m.fs.stage1.solvent_flux = pyo.Constraint(
+        m.fs.stage1.elements,
+        rule=solvent_rule,
+    )
+
+    m.fs.stage1.solute_sieving = pyo.Constraint(
+        m.fs.stage1.elements,
+        m.fs.solutes,
+        rule=solute_rule,
+    )
+
+    m.fs.stage1.retentate_inlet.flow_vol[0].fix(Q_feed)
+    m.fs.stage1.retentate_inlet.conc_mass_solute[0, "Li"].fix(C_Li_feed)
+    m.fs.stage1.retentate_inlet.conc_mass_solute[0, "Co"].fix(C_Co_feed)
+
+    assert degrees_of_freedom(m.fs.stage1) == 0
+
+    initializer = MSContactorInitializer()
```

Comment:
I would suggest separating the building of the model from the initialization of the model; i.e. have one function that builds the full model, and another that steps throguh initialization.

There might be rare cases you can;t make that separation, but I doubt this is one of them.